File: SOC/src/cc/game/soc/core/VertexType.java
package cc.game.soc.core;

import cc.game.soc.android.R;

public enum VertexType {
	OPEN(R.string.vertex_type_open, 0, false, false),
	
	SETTLEMENT(R.string.vertex_type_settlement, 0, false, true), // produce
	OPEN_SETTLEMENT(R.string.vertex_type_open_settlement, 0, false, true), // used to setup 'pirate island' games
	CITY(R.string.vertex_type_city, 0, false, true),
	WALLED_CITY(R.string.vertex_type_walled_city, 0, false, true),
	
	METROPOLIS_TRADE(R.string.vertex_type_metro_trade, 0, false, true),
	METROPOLIS_POLITICS(R.string.vertex_type_metro_politics, 0, false, true),
	METROPOLIS_SCIENCE(R.string.vertex_type_metro_science, 0, false, true),
	
	BASIC_KNIGHT_ACTIVE(R.string.vertex_type_knight_basic_active, 1, true, false),
	STRONG_KNIGHT_ACTIVE(R.string.vertex_type_knight_strong_active, 2, true, false),
	MIGHTY_KNIGHT_ACTIVE(R.string.vertex_type_knight_mighty_active, 3, true, false),

	BASIC_KNIGHT_INACTIVE(R.string.vertex_type_knight_basic_inactive, 1, false, false),
	STRONG_KNIGHT_INACTIVE(R.string.vertex_type_knight_strong_inactive, 2, false, false),
	MIGHTY_KNIGHT_INACTIVE(R.string.vertex_type_knight_mighty_inactive, 3, false, false),

	PIRATE_FORTRESS(R.string.vertex_type_pirate_fortress, 0, false, true),
	;
	
	VertexType(int nameId, int knightLevel, boolean active, boolean isStructure) {
		this.nameId = nameId;
		this.knightLevel = knightLevel;
		this.active = active;
		this.isStructure = isStructure;
	}
	
	final int nameId;
	final int knightLevel;
	final boolean active;
	final boolean isStructure;
	
	public int getKnightLevel() {
		return knightLevel;
	}
	
	public boolean isKnightActive() {
		return active;
	}
	
	public VertexType promotedType() {
		switch (this) {
			case BASIC_KNIGHT_ACTIVE:
				return STRONG_KNIGHT_ACTIVE;
			case BASIC_KNIGHT_INACTIVE:
				return STRONG_KNIGHT_INACTIVE;
			case STRONG_KNIGHT_ACTIVE:
				return MIGHTY_KNIGHT_ACTIVE;
			case STRONG_KNIGHT_INACTIVE:
				return MIGHTY_KNIGHT_INACTIVE;
			default:
				throw new SOCException("type not suitable to promote: " + name());
		}
	}

	public VertexType demotedType() {
		switch (this) {
			case STRONG_KNIGHT_ACTIVE:
				return BASIC_KNIGHT_ACTIVE;
			case STRONG_KNIGHT_INACTIVE:
				return BASIC_KNIGHT_INACTIVE;
			case MIGHTY_KNIGHT_ACTIVE:
				return STRONG_KNIGHT_ACTIVE;
			case MIGHTY_KNIGHT_INACTIVE:
				return STRONG_KNIGHT_INACTIVE;
			default:
				throw new SOCException("type not suitable to demote: " + name());
		}
	}

	public VertexType activatedType() {
		switch (this) {
			case BASIC_KNIGHT_INACTIVE:
				return BASIC_KNIGHT_ACTIVE;
			case MIGHTY_KNIGHT_INACTIVE:
				return MIGHTY_KNIGHT_ACTIVE;
			case STRONG_KNIGHT_INACTIVE:
				return STRONG_KNIGHT_ACTIVE;
			default:
				throw new SOCException("type not suitable to activeate: " + name());
		}
	}

	public VertexType deActivatedType() {
		switch (this) {
			case BASIC_KNIGHT_ACTIVE:
				return BASIC_KNIGHT_INACTIVE;
			case MIGHTY_KNIGHT_ACTIVE:
				return MIGHTY_KNIGHT_INACTIVE;
			case STRONG_KNIGHT_ACTIVE:
				return STRONG_KNIGHT_INACTIVE;
			default:
				throw new SOCException("type not suitable to deactiveate: " + name());
		}
	}

	public boolean isKnight() {
		switch (this) {
			case BASIC_KNIGHT_ACTIVE:
			case BASIC_KNIGHT_INACTIVE:
			case MIGHTY_KNIGHT_ACTIVE:
			case MIGHTY_KNIGHT_INACTIVE:
			case STRONG_KNIGHT_ACTIVE:
			case STRONG_KNIGHT_INACTIVE:
				return true;
			default:
				return false;
		}
	}

	public final String getName(StringResource sr) {
		return sr.getString(nameId);
	}
}

File: SOC/src/cc/game/soc/core/VertexType.kt
package cc.game.soc.core

enum class VertexType(val _nameId: String, val knightLevel: Int, val isKnightActive: Boolean, val isStructure: Boolean) {
	OPEN("Open", 0, false, false),
	SETTLEMENT("Settlement", 0, false, true),  // produce
	OPEN_SETTLEMENT("Open Settlement", 0, false, true),  // used to setup 'pirate island' games
	CITY("City", 0, false, true),
	WALLED_CITY("Walled City", 0, false, true),
	METROPOLIS_TRADE("Trade Metropolis", 0, false, true),
	METROPOLIS_POLITICS("Politics Metropolis", 0, false, true),
	METROPOLIS_SCIENCE("Science Metropolis", 0, false, true),
	BASIC_KNIGHT_ACTIVE("Basic Knight (A)", 1, true, false),
	STRONG_KNIGHT_ACTIVE("Strong Knight (A)", 2, true, false),
	MIGHTY_KNIGHT_ACTIVE("Mighty Knight (A)", 3, true, false),
	BASIC_KNIGHT_INACTIVE("Basic Knight (IA)", 1, false, false),
	STRONG_KNIGHT_INACTIVE("Strong Knight (IA)", 2, false, false),
	MIGHTY_KNIGHT_INACTIVE("Mighty Knight (IA)", 3, false, false),
	PIRATE_FORTRESS("Pirate Fortress", 0, false, true);

	fun promotedType(): VertexType {
		return when (this) {
			BASIC_KNIGHT_ACTIVE -> STRONG_KNIGHT_ACTIVE
			BASIC_KNIGHT_INACTIVE -> STRONG_KNIGHT_INACTIVE
			STRONG_KNIGHT_ACTIVE -> MIGHTY_KNIGHT_ACTIVE
			STRONG_KNIGHT_INACTIVE -> MIGHTY_KNIGHT_INACTIVE
			else                   -> throw SOCException("type not suitable to promote: $name")
		}
	}

	fun demotedType(): VertexType {
		return when (this) {
			STRONG_KNIGHT_ACTIVE -> BASIC_KNIGHT_ACTIVE
			STRONG_KNIGHT_INACTIVE -> BASIC_KNIGHT_INACTIVE
			MIGHTY_KNIGHT_ACTIVE -> STRONG_KNIGHT_ACTIVE
			MIGHTY_KNIGHT_INACTIVE -> STRONG_KNIGHT_INACTIVE
			else                   -> throw SOCException("type not suitable to demote: $name")
		}
	}

	fun activatedType(): VertexType {
		return when (this) {
			BASIC_KNIGHT_INACTIVE -> BASIC_KNIGHT_ACTIVE
			MIGHTY_KNIGHT_INACTIVE -> MIGHTY_KNIGHT_ACTIVE
			STRONG_KNIGHT_INACTIVE -> STRONG_KNIGHT_ACTIVE
			else                   -> throw SOCException("type not suitable to activeate: $name")
		}
	}

	fun deActivatedType(): VertexType {
		return when (this) {
			BASIC_KNIGHT_ACTIVE -> BASIC_KNIGHT_INACTIVE
			MIGHTY_KNIGHT_ACTIVE -> MIGHTY_KNIGHT_INACTIVE
			STRONG_KNIGHT_ACTIVE -> STRONG_KNIGHT_INACTIVE
			else                 -> throw SOCException("type not suitable to deactiveate: $name")
		}
	}

	val isKnight: Boolean
		get() = when (this) {
			BASIC_KNIGHT_ACTIVE, BASIC_KNIGHT_INACTIVE, MIGHTY_KNIGHT_ACTIVE, MIGHTY_KNIGHT_INACTIVE, STRONG_KNIGHT_ACTIVE, STRONG_KNIGHT_INACTIVE -> true
			else                                                                                                                                   -> false
		}

	val harborPts: Int
		get() = when (this) {
			SETTLEMENT -> 1
			CITY -> 2
			else -> 0
		}
}